# chick_NC-GRN

Here we provide additional information such as scripts and applications to access data generated in the chick_NC-GRN Project.


## Project Description

Precise control of developmental processes is encoded in the genome in the form of gene regulatory networks (GRNs). Such multi-factorial systems are difficult to decode in vertebrates owing to their complex gene hierarchies and transient dynamic molecular interactions. Here we present a genome-wide in vivo reconstruction of the GRN underlying development of neural crest (NC), an emblematic embryonic multipotent cell population. By coupling NC-specific epigenomic and single-cell transcriptome profiling with genome/epigenome engineering _in vivo_, we identify multiple regulatory layers governing NC ontogeny, including NC-specific enhancers and super-enhancers, novel trans-factors and _cis_-signatures. Assembling the NC regulome has allowed the comprehensive reverse engineering of the NC-GRN at unprecedented resolution. Furthermore, identification and dissection of divergent upstream combinatorial regulatory codes has afforded new insights into opposing gene circuits that define canonical and neural NC fates. Our integrated approach, allowing dissection of cell-type-specific regulatory circuits _in vivo_, has broad implications for GRN discovery and investigation.

### Keywords

Neural crest, gene regulatory network, chick, enhancers, ATAC, super-enhancers, Capture-C, transcription factors, neural crest cell lineage

### Research Organism

*Gallus gallus*



## Manuscript and contact information

**Reconstruction of the global neural crest gene regulatory network _in vivo_**

Ruth M Williams<sup>1</sup>, Ivan Candido-Ferreira<sup>1</sup>, Emmanouela Repapi<sup>2</sup>, Daria Gavriouchkina<sup>1,4</sup>, Upeka Senanayake<sup>1</sup>, Jelena Telenius<sup>2,3</sup>, Stephen Taylor<sup>2</sup>, Jim Hughes<sup>2,3</sup> and Tatjana Sauka-Spengler<sup>1,ยง</sup>.

 ยง Corresponding author: 
 **Tatjana Sauka-Spengler (tatjana.sauka-spengler@imm.ox.ac.uk)**

### Affiliations 

1. University of Oxford,
MRC Weatherall Institute of Molecular Medicine,
Radcliffe Department of Medicine,
Oxford, OX3 9DS, UK

2. University of Oxford,
MRC Centre for Computational Biology,
MRC Weatherall Institute of Molecular Medicine,
Oxford, OX3 9DS, UK

3. University of Oxford,
MRC Molecular Haematology Unit,
MRC Weatherall Institute of Molecular Medicine,
Oxford, OX3 9DS, UK

4. Present address:
Okinawa Institute of Science and Technology,
Molecular Genetics Unit,
Onna, 904-0495, Japan



Ruth M Williams, Ivan Candido-Ferreira, Daria Gavriouchkina, Upeka Senanayake, Tatjana Sauka-Spengler: [Sauka-Spengler lab](http://www.tsslab.co.uk) based at [MRC Weatherall Institute of Molecular Medicine](www.imm.ox.ac.uk) and [Radcliffe Department of Medicine](http://www.rdm.ox.ac.uk), University of Oxford.

Emmanouela Repapi, Stephen Taylor: [Taylor group](https://www.imm.ox.ac.uk/research/units-and-centres/mrc-wimm-centre-for-computational-biology/groups/taylor-group-visualisation-analysis-and-integration) based at [MRC Weatherall Institute of Molecular Medicine](http://www.imm.ox.ac.uk).

Jelena Telenius, Jim Hughes: [Hughes lab](https://www.imm.ox.ac.uk/research/units-and-centres/mrc-wimm-centre-for-computational-biology/groups/hughes-group) based at [MRC Weatherall Institute of Molecular Medicine](http://www.imm.ox.ac.uk) and [Radcliffe Department of Medicine](http://www.rdm.ox.ac.uk), University of Oxford.



### Acknowledgements

This work was supported by MRC (G0902418), The Lister Institute for Preventative Medicine Research Prize, John Fell Fund (131/038), and Leverhulme Trust grant (RPG-2015-026) to T.S.S. I.C.F is funded by the Oxford-Angus McLeod-St John's College Graduate fellowship and the WIMM prize studentship. 
